[WebDriver][BiDi] implement the browsingContext.handleUserPrompt command

burg opened the pull request:
#### 46a863cf3b36aba43a25ccbaf642c5e72ecf3666
<pre>
[WebDriver][BiDi] implement the browsingContext.handleUserPrompt command
<a href="https://bugs.webkit.org/show_bug.cgi?id=288328">https://bugs.webkit.org/show_bug.cgi?id=288328</a>
<a href="https://rdar.apple.com/145987385">rdar://145987385</a>

Reviewed by Devin Rousso.

Add basic implementations for handleUserPrompt, userPromptOpened, and userPromptClosed.
Add _WKAutomationSessionDelegate methods which allow the client to report the user-entered
text and default/placeholder text for the current dialog.

There are a few related pieces of work which need to happen separately as followups:

* Implement support for `File and `Beforeunload` user prompt types in WebKit.
<a href="https://bugs.webkit.org/show_bug.cgi?id=291665">https://bugs.webkit.org/show_bug.cgi?id=291665</a>

* Pass `unhandledPromptBehavior` from driver to WebKit. WebDriver Classic defines
an `unhandledPromptBehavior` capability which allows for automatically accepting, dismissing,
or ignoring user prompts during testing. WebDriver BiDi calls this the `user prompt handler`
and includes its value in the `browsingContext.userPromptOpened` event.
<a href="https://bugs.webkit.org/show_bug.cgi?id=291666">https://bugs.webkit.org/show_bug.cgi?id=291666</a>

* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiSession.json: Added.
The specification puts `UserPromptHandlerType` in the `session` domain since it is a
property of the session. Add a Session domain specification to keep the pattern going.

* Source/WebKit/UIProcess/API/APIAutomationSessionClient.h:
(API::AutomationSessionClient::defaultTextOfCurrentJavaScriptDialogOnPage): Added.
(API::AutomationSessionClient::userInputOfCurrentJavaScriptDialogOnPage): Added.
These are needed by `browsingContext.userPromptOpened` and `browsingContext.userPromptClosed`.

* Source/WebKit/UIProcess/API/Cocoa/_WKAutomationSessionDelegate.h:
* Source/WebKit/UIProcess/Cocoa/AutomationSessionClient.h:
* Source/WebKit/UIProcess/Cocoa/AutomationSessionClient.mm:
(WebKit::AutomationSessionClient::AutomationSessionClient):
(WebKit::AutomationSessionClient::defaultTextOfCurrentJavaScriptDialogOnPage):
(WebKit::AutomationSessionClient::userInputOfCurrentJavaScriptDialogOnPage):
Add delegate methods and boilerplate to match new APIAutomationSessionClient methods.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::toProtocolUserPromptType): Added.
(WebKit::WebAutomationSession::willShowJavaScriptDialog):
Emit userPromptOpened after the run loop spins once and the dialog has definitely appeared.

(WebKit::WebAutomationSession::dismissCurrentJavaScriptDialog):
(WebKit::WebAutomationSession::acceptCurrentJavaScriptDialog):
Emit userPromptClosed after accepting or dismissing the user prompt.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
Drive-by, remove unnecessary forward declaration for a class that was renamed.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
Add frontend dispatcher object for `browsingContext` domain.

(WebKit::WebDriverBidiProcessor::handleUserPrompt):
Added. Accept or dismiss the user prompt.

(WebKit::WebDriverBidiProcessor::userPromptOpenedOnPage):
(WebKit::WebDriverBidiProcessor::userPromptClosedOnPage):
Added. These methods are pass through, the data is fetched from the client in WebAutomationSession.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
Add BidiSession.json.

Canonical link: <a href="https://commits.webkit.org/293978@main">https://commits.webkit.org/293978@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63d558cf9330c6f524614a1bbeeadee8d58324cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100445 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20097 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10396 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105582 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51033 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20404 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28571 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76484 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33536 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103452 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15636 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90731 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56840 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15452 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50408 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85369 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8813 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107936 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27563 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20229 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85436 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27926 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86931 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84974 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21621 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29659 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7390 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21528 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27498 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32748 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27309 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30627 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28867 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -719,6 +741,17 @@ void WebAutomationSession::willShowJavaScriptDialog(WebPageProxy& page)
+        String dialogMessage = m_client->messageOfCurrentJavaScriptDialogOnPage(*this, page);
+        String dialogDefaultText = m_client->defaultTextOfCurrentJavaScriptDialogOnPage(*this, page);
+        std::optional<API::AutomationSessionClient::JavaScriptDialogType> apiDialogType = m_client->typeOfCurrentJavaScriptDialogOnPage(*this, page);
+        Inspector::Protocol::BidiBrowsingContext::UserPromptType userPromptType = toProtocolUserPromptType(apiDialogType.value_or(API::AutomationSessionClient::JavaScriptDialogType::Prompt));
```
NIT: i feel like both of these could/should be `auto`

burg replied:
I'll try that again, I recall it not being able to infer the optional type at least.

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -1241,6 +1274,12 @@ CommandResult<void> WebAutomationSession::dismissCurrentJavaScriptDialog(const I
 
+#if ENABLE(WEBDRIVER_BIDI)
+    auto apiDialogType = m_client->typeOfCurrentJavaScriptDialogOnPage(*this, *page);
+    SYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!apiDialogType, InternalError);
```
NIT: should this still fail even if the client is not using WebDriver BiDi (i.e. WebDriver "classic")?

dcrousso commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -1257,6 +1296,14 @@ CommandResult<void> WebAutomationSession::acceptCurrentJavaScriptDialog(const In
 
+#if ENABLE(WEBDRIVER_BIDI)
+    auto apiDialogType = m_client->typeOfCurrentJavaScriptDialogOnPage(*this, *page);
+    SYNC_FAIL_WITH_PREDEFINED_ERROR_IF(!apiDialogType, InternalError);
```
ditto

yury-s commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -719,6 +741,17 @@ void WebAutomationSession::willShowJavaScriptDialog(WebPageProxy& page)
             return;
 
+#if ENABLE(WEBDRIVER_BIDI)
+        String dialogMessage = m_client->messageOfCurrentJavaScriptDialogOnPage(*this, page);
```
Both of these values (`message` and `defaultText`) are already available in [WebPageProxy](https://github.com/WebKit/WebKit/blob/f2f4f6420f2124e158906e800e84035a6ae9f46b/Source/WebKit/UIProcess/WebPageProxy.cpp#L8717) where `willShowJavaScriptDialog` is called. You can just pass them as parameters to this method and avoid all the embedder plumbing. Would that work?

burg replied:
Thanks for the tip. I think they would be good fallback values rather than failing as Devin also pointed out. We do need to ask the embedder because it may apply practical limits on string lengths.

yury-s replied:
I don't see the Devin's suggestion that you refer to in the review, but using those values at least as a fallback / default implementation where embedder lacks more specific one, would be nice.

We've been returning the values that WebKit passes to the embedder in Playwright, and over the years haven't seen a single complaint from the users about the text length limits being not respected, so I'm not sure if it's even an issue in practice.

burg committed:
[WebDriver][BiDi] implement the browsingContext.handleUserPrompt command
https://bugs.webkit.org/show_bug.cgi?id=288328
rdar://145987385

Reviewed by Devin Rousso.

Add basic implementations for handleUserPrompt, userPromptOpened, and userPromptClosed.
Add _WKAutomationSessionDelegate methods which allow the client to report the user-entered
text and default/placeholder text for the current dialog.

There are a few related pieces of work which need to happen separately as followups:

* Implement support for `File and `Beforeunload` user prompt types in WebKit.
https://bugs.webkit.org/show_bug.cgi?id=291665

* Pass `unhandledPromptBehavior` from driver to WebKit. WebDriver Classic defines
an `unhandledPromptBehavior` capability which allows for automatically accepting, dismissing,
or ignoring user prompts during testing. WebDriver BiDi calls this the `user prompt handler`
and includes its value in the `browsingContext.userPromptOpened` event.
https://bugs.webkit.org/show_bug.cgi?id=291666

* Source/WebKit/UIProcess/Automation/protocol/BidiBrowsingContext.json:
* Source/WebKit/UIProcess/Automation/protocol/BidiSession.json: Added.
The specification puts `UserPromptHandlerType` in the `session` domain since it is a
property of the session. Add a Session domain specification to keep the pattern going.

* Source/WebKit/UIProcess/API/APIAutomationSessionClient.h:
(API::AutomationSessionClient::defaultTextOfCurrentJavaScriptDialogOnPage): Added.
(API::AutomationSessionClient::userInputOfCurrentJavaScriptDialogOnPage): Added.
These are needed by `browsingContext.userPromptOpened` and `browsingContext.userPromptClosed`.

* Source/WebKit/UIProcess/API/Cocoa/_WKAutomationSessionDelegate.h:
* Source/WebKit/UIProcess/Cocoa/AutomationSessionClient.h:
* Source/WebKit/UIProcess/Cocoa/AutomationSessionClient.mm:
(WebKit::AutomationSessionClient::AutomationSessionClient):
(WebKit::AutomationSessionClient::defaultTextOfCurrentJavaScriptDialogOnPage):
(WebKit::AutomationSessionClient::userInputOfCurrentJavaScriptDialogOnPage):
Add delegate methods and boilerplate to match new APIAutomationSessionClient methods.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::toProtocolUserPromptType): Added.
(WebKit::WebAutomationSession::willShowJavaScriptDialog):
Emit userPromptOpened after the run loop spins once and the dialog has definitely appeared.

(WebKit::WebAutomationSession::dismissCurrentJavaScriptDialog):
(WebKit::WebAutomationSession::acceptCurrentJavaScriptDialog):
Emit userPromptClosed after accepting or dismissing the user prompt.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
Drive-by, remove unnecessary forward declaration for a class that was renamed.

* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
Add frontend dispatcher object for `browsingContext` domain.

(WebKit::WebDriverBidiProcessor::handleUserPrompt):
Added. Accept or dismiss the user prompt.

(WebKit::WebDriverBidiProcessor::userPromptOpenedOnPage):
(WebKit::WebDriverBidiProcessor::userPromptClosedOnPage):
Added. These methods are pass through, the data is fetched from the client in WebAutomationSession.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
Add BidiSession.json.

Canonical link: https://commits.webkit.org/293978@main
